### PR TITLE
Software types can be used in groovy scripts without plugins block

### PR DESCRIPTION
**Pull request by ghale**
Allow groovy scripts to reference and apply software types without an explicit `plugins { }` block.

This also pulls plugin application fully into the action of software feature application.  In other words, now that all script types can implicitly reference software types, applying the software type implicitly applies its associated plugin.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by ghale** [comment]
@bot-gradle test this

**Review by eskatos — APPROVED** [review]
💯

File: subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java [hunk 1]
```diff
@@ -257,13 +260,15 @@ public DefaultProject(
...
         );
 
-
         @Nullable DynamicObject parentInherited = services.get(CrossProjectModelAccess.class).parentProjectDynamicInheritedScope(this);
         if (parentInherited != null) {
             extensibleDynamicObject.setParent(parentInherited);
         }
         extensibleDynamicObject.addObject(taskContainer.getTasksAsDynamicObject(), ExtensibleDynamicObject.Location.AfterConvention);
 
+        DynamicObject softwareFeaturesDynamicObject = new SoftwareFeaturesDynamicObject(getSoftwareTypeRegistry(), getSoftwareFeatureApplicator(), this);
```
Comment (tresat) [inline subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProject.java:269]: 💭 Should the `SoftwareFeaturesDynamicObject` itself be the project-scoped service?  There there would be only one new getter to inject into `DefaultProject`, which would be constructed with the appropriate project instance in the service registry.

**Review by tresat — APPROVED** [review]

File: subprojects/core/src/main/java/org/gradle/plugin/software/internal/SoftwareFeaturesDynamicObject.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+import groovy.lang.Closure;
+import org.gradle.api.plugins.ExtensionAware;
+import org.gradle.internal.metaobject.AbstractDynamicObject;
+import org.gradle.internal.metaobject.DynamicInvokeResult;
+import org.gradle.util.internal.ConfigureUtil;
+
+import javax.annotation.Nullable;
+
+public class SoftwareFeaturesDynamicObject extends AbstractDynamicObject {
```
[the remark this answers is not in the document]
Comment (ghale) [reply subprojects/core/src/main/java/org/gradle/plugin/software/internal/SoftwareFeaturesDynamicObject.java:32]: Done.

**Pull request merged**